llama_index/indices/query/query_transform/feedback_transform.py

```diff
@@ -1,37 +1,105 @@
+    "{feedback}\n"
+    "If you want to resynthesize the query, please return the modified query below.\n"
+    "Otherwise, please return the original query.\n"
+)
+
+DEFAULT_RESYNTHESIS_PROMPT = Prompt(DEFAULT_RESYNTHESIS_PROMPT_TMPL)
 
 
 class FeedbackQueryTransformation(BaseQueryTransform):
     """Transform the query given the evaluation feedback.
 
     Args:
         eval(Evaluation): An evaluation object.
+        llm_predictor(BaseLLMPredictor): An LLM predictor.
+        resynthesize_query(bool): Whether to resynthesize the query.
+        resynthesis_prompt(Prompt): A prompt for resynthesizing the query.
 
     """
 
     def __init__(
         self,
-        evaluation: Evaluation,
+        evaluation: Optional[Evaluation] = None,
+        llm_predictor: Optional[BaseLLMPredictor] = None,
+        resynthesize_query: bool = False,
+        resynthesis_prompt: Optional[Prompt] = None,
     ) -> None:
         super().__init__()
         self.evaluation = evaluation
+        self.llm_predictor = llm_predictor or LLMPredictor()
+        self.should_resynthesize_query = resynthesize_query
+        self.resynthesis_prompt = resynthesis_prompt or DEFAULT_RESYNTHESIS_PROMPT
+
+    def set_eval(self, evaluation: Evaluation) -> None:
+        """Set evaluation."""
+        self.evaluation = evaluation
 
     def _run(self, query_bundle: QueryBundle, extra_info: Dict) -> QueryBundle:
-        query_str = query_bundle.query_str
-        new_query = (
-            query_str
-            + "\n----------------\n"
-            + self.construct_feedback(self.evaluation.response.response)
-        )
-        return QueryBundle(new_query)
-
-    def construct_feedback(self, response: Optional[str]) -> str:
+        orig_query_str = query_bundle.query_str
+        if self.evaluation is None:
+            raise ValueError("Evaluation is not set.")
+        if self.evaluation.feedback == "YES" or self.evaluation.feedback == "NO":
+            new_query = (
+                orig_query_str
+                + "\n----------------\n"
+                + self._construct_feedback(response=self.evaluation.response.response)
+            )
+        else:
+            if self.should_resynthesize_query:
+                new_query_str = self._resynthesize_query(
+                    orig_query_str, self.evaluation.response, self.evaluation.feedback
+                )
+            else:
+                new_query_str = orig_query_str
+            new_query = (
+                self._construct_feedback(response=self.evaluation.response.response)
+                + "\n"
+                + "Here is some feedback from the evaluator about the response given.\n"
+                + self.evaluation.feedback
+                + "\n"
+                + "Now answer the question.\n"
+                + new_query_str
+            )
+        return QueryBundle(new_query, custom_embedding_strs=[orig_query_str])
+
+    def _construct_feedback(self, response: Optional[str]) -> str:
```

Comment:
very nit: this really doesn't need to be an instance method.

Reply:
Fixed